pkg/scheduler/schedule_one.go

```diff
@@ -711,40 +711,31 @@ func prioritizeNodes(
 
 	if len(extenders) != 0 && nodes != nil {
 		var mu sync.Mutex
-		var wg sync.WaitGroup
 		combinedScores := make(map[string]int64, len(nodes))
-		for i := range extenders {
-			if !extenders[i].IsInterested(pod) {
-				continue
+		fwk.Parallelizer().Until(ctx, len(extenders), func(extIndex int) {
```

Comment:
What can we benefit from this?

Reply:
parallelizer allows us to limit the number of goroutines specified by [the `WithParallelism ` option](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/scheduler.go#L158)

Reply:
I guess for most cases, the number of extenders are smaller than the default parallelism which is defined as 16 today.

Reply:
Yes, I too think almost all users use extenders much smaller than 16.
But, users can set any value with [WithParallelism](https://github.com/kubernetes/kubernetes/blob/1d2e8042877c4facd3a45e911857f92474d64797/pkg/scheduler/scheduler.go#L169) option as I said before. 
We need to care about it in any goroutine creation in scheduling loop because that's what we promise our users in `WithParallelism` option.

Also, the benefit we have here is removing the `goroutine` (also `scheduler_goroutine`) metric increment here (as parallelizer increments the metric internally.).
The motivation is the similar to ↓. 
> It makes me think that we are not very diligent incrementing these counters throughout the codebase for everytime we create routines. 
https://github.com/kubernetes/kubernetes/pull/111775#discussion_r944671465

---

@denkensk 

> I am not sure if it will cause performance degradation when change to use Parallelizer. As you know, Parallelizer will do extra things to prepare.

True. I checked just in case. 
And, actually, no performance degradation when the number of extenders = 1.
Also checked when the number of extenders = 20 (more than 16), we can see the performance degradation, of course, as the parallelizer limits the number of extenders which run at the same time.

<details><summary>The benchmark test detail</summary><div>

　

**The benchmark implementation**

(It's something like https://github.com/kubernetes/kubernetes/pull/112220.)

```go

func BenchmarkSchedulePod(b *testing.B) {
	extenders := make([]framework.Extender, 20)
	for i := 0; i < 20; i++ {
		extenderName := "FakeExtender" + strconv.Itoa(i)
		extenders[i] = &st.FakeExtender{
			ExtenderName: extenderName,
			Predicates:   []st.FitPredicate{st.TruePredicateExtender},
			Prioritizers: []st.PriorityConfig{{Function: st.Node1PrioritizerExtender, Weight: 10}},
			Weight:       1,
		}
	}

	nodes := make([]string, 100)
	for i := 0; i < 100; i++ {
		nodes[i] = strconv.Itoa(i)
	}
	tests := []struct {
		name            string
		registerPlugins []st.RegisterPluginFunc
		extenders       []framework.Extender
		nodes           []string
		pvcs            []v1.PersistentVolumeClaim
		pod             *v1.Pod
		pods            []*v1.Pod
	}{
		{
			name: "Scheduler has 20 extenders, but has no plugins.",
			registerPlugins: []st.RegisterPluginFunc{
				st.RegisterQueueSortPlugin(queuesort.Name, queuesort.New),
				st.RegisterBindPlugin(defaultbinder.Name, defaultbinder.New),
			},
			extenders: extenders,
			nodes:     nodes,
			pod:       st.MakePod().Name("pod").Obj(),
		},
		{
			name: "Scheduler has 1 extender, but has no plugins.",
			registerPlugins: []st.RegisterPluginFunc{
				st.RegisterQueueSortPlugin(queuesort.Name, queuesort.New),
				st.RegisterBindPlugin(defaultbinder.Name, defaultbinder.New),
			},
			extenders: extenders[:1],
			nodes:     nodes,
			pod:       st.MakePod().Name("pod").Obj(),
		},
	}

	for _, test := range tests {
		b.Run(test.name, func(b *testing.B) {
			cache := internalcache.New(time.Duration(0), wait.NeverStop)
			for _, pod := range test.pods {
				cache.AddPod(pod)
			}
			var nodes []*v1.Node
			for _, name := range test.nodes {
				node := &v1.Node{ObjectMeta: metav1.ObjectMeta{Name: name, Labels: map[string]string{"hostname": name}}}
				nodes = append(nodes, node)
				cache.AddNode(node)
			}

			ctx, cancel := context.WithCancel(context.Background())
			defer cancel()
			cs := clientsetfake.NewSimpleClientset()
			informerFactory := informers.NewSharedInformerFactory(cs, 0)
			for _, pvc := range test.pvcs {
				metav1.SetMetaDataAnnotation(&pvc.ObjectMeta, volume.AnnBindCompleted, "true")
				cs.CoreV1().PersistentVolumeClaims(pvc.Namespace).Create(ctx, &pvc, metav1.CreateOptions{})
				if pvName := pvc.Spec.VolumeName; pvName != "" {
					pv := v1.PersistentVolume{ObjectMeta: metav1.ObjectMeta{Name: pvName}}
					cs.CoreV1().PersistentVolumes().Create(ctx, &pv, metav1.CreateOptions{})
				}
			}
			snapshot := internalcache.NewSnapshot(test.pods, nodes)
			fwk, err := st.NewFramework(
				test.registerPlugins, "", ctx.Done(),
				frameworkruntime.WithSnapshotSharedLister(snapshot),
				frameworkruntime.WithInformerFactory(informerFactory),
				frameworkruntime.WithPodNominator(internalqueue.NewPodNominator(informerFactory.Core().V1().Pods().Lister())),
			)
			if err != nil {
				b.Fatal(err)
			}

			scheduler := newScheduler(
				cache,
				test.extenders,
				nil,
				nil,
				nil,
				nil,
				nil,
				snapshot,
				schedulerapi.DefaultPercentageOfNodesToScore)
			informerFactory.Start(ctx.Done())
			informerFactory.WaitForCacheSync(ctx.Done())

			for i := 0; i < b.N; i++ {
				_, err := scheduler.SchedulePod(ctx, fwk, framework.NewCycleState(), test.pod)
				if err != nil {
					b.Fatal(err)
				}
			}
		})
	}
}
```


**result (this branch)**

```
Running tool: /home/gitpod/go/bin/go test -benchmem -run=^$ -bench ^BenchmarkSchedulePod$ k8s.io/kubernetes/pkg/scheduler

goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/scheduler
cpu: AMD EPYC 7B13
BenchmarkSchedulePod/Scheduler_has_20_extenders,_but_has_no_plugins.-16         	    1270	    906839 ns/op	  717101 B/op	    5072 allocs/op
BenchmarkSchedulePod/Scheduler_has_1_extender,_but_has_no_plugins.-16           	    9300	    123879 ns/op	   49111 B/op	     502 allocs/op
PASS
ok  	k8s.io/kubernetes/pkg/scheduler	6.524s
```

**result (master)**

```
Running tool: /home/gitpod/go/bin/go test -benchmem -run=^$ -bench ^BenchmarkSchedulePod$ k8s.io/kubernetes/pkg/scheduler

goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/scheduler
cpu: AMD EPYC 7B13
BenchmarkSchedulePod/Scheduler_has_20_extenders,_but_has_no_plugins.-16         	     450	   2569347 ns/op	  719380 B/op	    5101 allocs/op
BenchmarkSchedulePod/Scheduler_has_1_extender,_but_has_no_plugins.-16           	    9426	    122577 ns/op	   48915 B/op	     497 allocs/op
PASS
ok  	k8s.io/kubernetes/pkg/scheduler	8.814s
```

</div></details>

But, I'm wondering if we could improve the parallelizer to not do extra thing (like starting `workqueue`) 
when parallelism > pieces given to `parallelizer.Until`.